flepimop/gempyor_pkg/tests/parameters/test_parameters_class.py

```diff
@@ -0,0 +1,508 @@
+            assert params.pconfig == valid_parameters
+
+            # The `pdata` attribute
+            assert set(params.pdata.keys()) == {"sigma", "gamma", "Ro"}
+            assert set(params.pdata["sigma"].keys()) == {
+                "idx",
+                "ts",
+                "stacked_modifier_method",
+            }
+            assert params.pdata["sigma"]["idx"] == 0
+            assert params.pdata["sigma"]["ts"].equals(param_df.set_index("date"))
+            assert params.pdata["sigma"]["stacked_modifier_method"] == "product"
+            assert set(params.pdata["gamma"].keys()) == {
+                "idx",
+                "dist",
+                "stacked_modifier_method",
+            }
+            assert params.pdata["gamma"]["idx"] == 1
+            assert isinstance(params.pdata["gamma"]["dist"], partial)
+            assert partials_are_similar(
+                params.pdata["gamma"]["dist"],
+                partial(np.random.uniform, 0.1234, 0.1234),
+            )
+            assert params.pdata["gamma"]["stacked_modifier_method"] == "sum"
+            assert set(params.pdata["Ro"].keys()) == {
+                "idx",
+                "dist",
+                "stacked_modifier_method",
+            }
+            assert params.pdata["Ro"]["idx"] == 2
+            assert isinstance(params.pdata["Ro"]["dist"], partial)
+            assert partials_are_similar(
+                params.pdata["Ro"]["dist"], partial(np.random.uniform, 1.0, 2.0)
+            )
+            assert params.pdata["Ro"]["stacked_modifier_method"] == "product"
+
+            # The `pnames` attribute
+            assert params.pnames == ["sigma", "gamma", "Ro"]
+
+            # The `pnames2pindex` attribute
+            assert params.pnames2pindex == {"sigma": 0, "gamma": 1, "Ro": 2}
+
+            # The `stacked_modifier_method` attribute
+            assert params.stacked_modifier_method == {
+                "sum": ["gamma"],
+                "product": ["sigma", "ro"],
+                "reduction_product": [],
+            }
+
+    def test_picklable_lamda_alpha(self) -> None:
+        # Setup
+        simple_parameters = create_confuse_subview_from_dict(
+            "parameters", {"sigma": {"value": 0.1}}
+        )
+        params = Parameters(
+            simple_parameters,
+            ti=date(2024, 1, 1),
+            tf=date(2024, 1, 10),
+            subpop_names=["1", "2"],
+        )
+
+        # Attribute error if `alpha_val` is not set
+        with pytest.raises(AttributeError):
+            params.picklable_lamda_alpha()
+
+        # We get the expected value when `alpha_val` is set
+        params.alpha_val = None
+        assert params.picklable_lamda_alpha() == None
+
+    def test_picklable_lamda_sigma(self) -> None:
+        # Setup
+        simple_parameters = create_confuse_subview_from_dict(
+            "parameters", {"sigma": {"value": 0.1}}
+        )
+        params = Parameters(
```

Comment:
seems like there are some standard comparison objects defined (e.g. i see this appears in lines above) - presumably pytest let's you define them once somewhere, and then re-use in multiple tests?

Reply:
I'm not exactly sure what this means in this context? Do mean like defining `simple_parameters` globally somewhere? We can either define those at the class level or at the top of the file, not sure of the pros/cons of either. I don't know enough about the pytest runner to know if declaring them at the top of the file clutters global environment or if pytest isolates files, easy to find out though.

Reply:
Yes, "globally" (in the test context) declaring a standard set of comparison / example objects / inputs / etc.